### PR TITLE
Connection level bulk copy metadata caching

### PR DESCRIPTION
**Pull request by tkyc**
- Added connection level cache for destination column metadata.
- Added `cacheBulkCopyMetadata` boolean connection property.
  - true: driver will cache the metadata at the connection level
  - false: driver will not do any connection level metadata caching

Caveat: If the table changes in any way (column definition, new columns, deleted columns, etc..), then during a bulk copy the operation may or may not error out when caching is enabled. Just a note that there could be unintended consequences without warning.

### CODE REVIEW

**Review by barryw-mssql — DISMISSED** [review]

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerBulkCopy.java [hunk 1]
```diff
@@ -3709,4 +3752,18 @@ private boolean unicodeConversionRequired(int jdbcType, SSType ssType) {
...
+            for (String secret : secrets) {
+                if (null != secret) {
+                    md.update(secret.getBytes(java.nio.charset.StandardCharsets.UTF_16LE));
+                }
+            }
+            return new String(md.digest());
+        } catch (NoSuchAlgorithmException e) {
+            throw new SQLServerException(SQLServerException.getErrString("R_NoSHA256Algorithm"), e);
+        }
+    }
```
Comment (lilgreenbird) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerBulkCopy.java:3768]: this is a dupe of the same method in SQLServerMSAL4JUtils.java can we refactor that into a common place

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerBulkCopy.java [hunk 2]
```diff
@@ -1719,61 +1729,94 @@ private void getDestinationMetadata() throws SQLServerException {
...
+
+        if (connection.getEnableBulkCopyCache()) {
+            String databaseName = connection.activeConnectionProperties
+                    .getProperty(SQLServerDriverStringProperty.DATABASE_NAME.toString());
+            key = getHashedSecret(new String[] {escapedDestinationTableName, databaseName});
+            destColumnMetadata = BULK_COPY_OPERATION_CACHE.get(key);
+        }
+
+        if (null == destColumnMetadata || destColumnMetadata.isEmpty()) {
+
```
Comment (lilgreenbird) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerBulkCopy.java:1743]: nitpick: extra blank line here

File: src/main/java/com/microsoft/sqlserver/jdbc/ISQLServerConnection.java [hunk 3]
```diff
@@ -532,4 +532,19 @@ CallableStatement prepareCall(String sql, int nType, int nConcur, int nHold,
      * @return flag for using Bulk Copy API for batch insert operations.
      */
     boolean getUseBulkCopyForBatchInsert();
+
+    /**
+     * Returns value of 'enableBulkCopyCache' from Connection String.
+     *
+     * @param enableBulkCopyCache
```
Comment (David-Engel) [inline src/main/java/com/microsoft/sqlserver/jdbc/ISQLServerConnection.java:539]: I would call the option `cacheBulkCopyMetadata`
```suggestion
     * @param cacheBulkCopyMetadata
```

**Review by barryw-mssql — DISMISSED** [review]

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerBulkCopy.java [hunk 4]
```diff
@@ -1719,61 +1730,104 @@ private void getDestinationMetadata() throws SQLServerException {
...
+                        // defeat the purpose of caching.
+                        BULK_COPY_OPERATION_CACHE.put(key, destColumnMetadata);
+                    } finally {
+                        DESTINATION_COL_METADATA_LOCK.unlock();
+                    }
+                }
+            } else {
+                setDestinationColumnMetadata(escapedDestinationTableName);
+            }
+        }
```
Comment (David-Engel) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerBulkCopy.java:1769]: Add logging. Suggest FINER level and log if the metadata came from the cache or if it was pulled from the DB.

**Review by barryw-mssql — APPROVED** [review]

**Review by Jeffery-Wasty — APPROVED** [review]

**Comment by Jeffery-Wasty** [comment]
@microsoft-github-policy-service agree

**Comment by Jeffery-Wasty** [comment]
@microsoft-github-policy-service agree company="Microsoft"

**Review by Jeffery-Wasty — DISMISSED** [review]

**Review by barryw-mssql — DISMISSED** [review]

**Review by Jeffery-Wasty — APPROVED** [review]

**Review by Jeffery-Wasty — DISMISSED** [review]

**Review by barryw-mssql — APPROVED** [review]

**Review by Jeffery-Wasty — APPROVED** [review]

**Review by barryw-mssql — APPROVED** [review]

**Pull request merged**